frappe/integration_broker/oauth2.py

```diff
@@ -0,0 +1,139 @@
+from __future__ import unicode_literals
+import frappe, json
+from frappe.oauth import OAuthWebRequestValidator, WebApplicationServer
+from oauthlib.oauth2 import FatalClientError, OAuth2Error
+from urllib import quote, urlencode
+from frappe.integrations.doctype.oauth_provider_settings.oauth_provider_settings import get_oauth_settings
+
+#Variables required across requests
+oauth_validator = OAuthWebRequestValidator()
+oauth_server  = WebApplicationServer(oauth_validator)
+credentials = None
+
+def get_urlparams_from_kwargs(param_kwargs):
+	arguments = param_kwargs
+	if arguments.get("data"):
+		arguments.pop("data")
+	if arguments.get("cmd"):
+		arguments.pop("cmd")
+
+	return urlencode(arguments)
+
+@frappe.whitelist()
+def approve(*args, **kwargs):
+	r = frappe.request
+	uri = r.url
+	http_method = r.method
+	body = r.get_data()
+	headers = r.headers
+
+	try:
+		scopes, credentials = oauth_server.validate_authorization_request(uri, http_method, body, headers)
+
+		headers, body, status = oauth_server.create_authorization_response(uri=credentials['redirect_uri'], \
+				body=body, headers=headers, scopes=scopes, credentials=credentials)
+		uri = headers.get('Location', None)
+
+		frappe.local.response["type"] = "redirect"
+		frappe.local.response["location"] = uri
+
+	except FatalClientError as e:
+		return e
+	except OAuth2Error as e:
+		return e
+
+@frappe.whitelist(allow_guest=True)
+def authorize(*args, **kwargs):
+	#Fetch provider URL from settings
+	oauth_settings = get_oauth_settings()
+	params = get_urlparams_from_kwargs(kwargs)
+	success_url = oauth_settings["provider_url"] + "/api/method/frappe.integration_broker.oauth2.approve?" + params
+
+	if frappe.session['user']=='Guest':
+		#Force login, redirect to preauth again.
+		frappe.local.response["type"] = "redirect"
+		frappe.local.response["location"] = "/login?redirect-to=/api/method/frappe.integration_broker.oauth2.authorize?" + quote(params)
+
+	elif frappe.session['user']!='Guest':
+		try:
+			r = frappe.request
+			uri = r.url
+			http_method = r.method
+			body = r.get_data()
+			headers = r.headers
+
+			scopes, credentials = oauth_server.validate_authorization_request(uri, http_method, body, headers)
+
+			skipauth = frappe.db.get_value("OAuth Client", credentials['client_id'], "skip_authorization")
```

Comment:
Maybe skip_auth instead of skipauth. 

Skipauth sounds like a nice band name though.